In [173]:
# Import libraries necessary for this project
import numpy as np
import pandas as pd
from sklearn.cross_validation import ShuffleSplit

# Import supplementary visualizations code visuals.py
import visuals as vs

# Pretty display for notebooks
%matplotlib inline

# Load the Boston housing dataset
data = pd.read_csv('housing.csv')
prices = data['MEDV']
features = data.drop('MEDV', axis = 1)
#np.asarray(prices)
# Success
print ("Boston housing dataset has {} data points with {} variables each.".format(*data.shape))

Boston housing dataset has 489 data points with 4 variables each.


In [174]:
# Minimum price of the data
minimum_price = min(prices)

# TODO: Maximum price of the data
maximum_price = max(prices)

#  Mean price of the data
mean_price =  np.mean(prices)

#  Median price of the data
median_price = np.median (prices)

# Standard deviation of prices of the data
std_price = np.std(prices)

# Show the calculated statistics
print ("Statistics for Boston housing dataset:\n")
print ("Minimum price: ${:,.2f}".format(minimum_price))
print ("Maximum price: ${:,.2f}".format(maximum_price))
print ("Mean price: ${:,.2f}".format(mean_price))
print ("Median price ${:,.2f}".format(median_price))
print ("Standard deviation of prices: ${:,.2f}".format(std_price))

Statistics for Boston housing dataset:

Minimum price: $105,000.00
Maximum price: $1,024,800.00
Mean price: $454,342.94
Median price $438,900.00
Standard deviation of prices: $165,171.13



Question 1 - Feature Observation
As a reminder, we are using three features from the Boston housing dataset: 'RM', 'LSTAT', and 'PTRATIO'. For each data point (neighborhood):

'RM' is the average number of rooms among homes in the neighborhood.
'LSTAT' is the percentage of homeowners in the neighborhood considered "lower class" (working poor).
'PTRATIO' is the ratio of students to teachers in primary and secondary schools in the neighborhood.
Using your intuition, for each of the three features above, do you think that an increase in the value of that feature would lead to an increase in the value of 'MEDV' or a decrease in the value of 'MEDV'? Justify your answer for each.

Hint: This problem can phrased using examples like below.

Would you expect a home that has an 'RM' value(number of rooms) of 6 be worth more or less than a home that has an 'RM' value of 7?

In [175]:
#slicing prices acrroding to number of rooms
# get the mean price of house if number of room < 6
print(round(np.mean(prices [features['RM'] < 6])))
# get the mean price of house if number of room > 6
print(round(np.mean(prices [features['RM'] > 6])))
# for values of RM increasing the prices increse also.

358830
505704


In [172]:
# Round RM values to make more sense 
# just for check
RoundedRM=round(features['RM'])
print(round(np.mean(prices [RoundedRM == 4])))
print(round(np.mean(prices [RoundedRM == 5])))
print(round(np.mean(prices [RoundedRM == 6])))
print(round(np.mean(prices [RoundedRM == 7])))
print(round(np.mean(prices [RoundedRM == 8])))

357420
293417
402536
577655
910219


Would you expect a neighborhood that has an 'LSTAT' value(percent of lower class workers) of 15 have home prices be worth more or less than a neighborhood that has an 'LSTAT' value of 20?

In [176]:
print(round(np.mean(prices [features['LSTAT'] < 15])))
print(round(np.mean(prices [features['LSTAT'] > 15])))
print(round(np.mean(prices [features['LSTAT'] > 20])))
# for LSTAT values increasing the prices decrease

527325
307028
259918


Would you expect a neighborhood that has an 'PTRATIO' value(ratio of students to teachers) of 10 have home prices be worth more or less than a neighborhood that has an 'PTRATIO' value of 15?

In [177]:
print(round(np.mean(prices [features['PTRATIO'] < 15])))
print(round(np.mean(prices [features['PTRATIO'] > 15])))
# for PTRATIO values increasing the prices decrease

564471
442079


## Question 1 ansewer 

 Feature Observation 
* an increase in the value of RM  feature would lead to an increase in the value of 'MEDV' .
* an increase in the value of RM  feature would lead to an decrease in the value of 'MEDV' .
* an increase in the value of RM  feature would lead to an decrease in the value of 'MEDV' .

In [184]:
# Import 'r2_score' coefficient of determination R^2
from sklearn.metrics import r2_score

def performance_metric(y_true, y_predict):
    """ Calculates and returns the performance score between 
        true and predicted values based on the metric chosen. """
    
# Calculate the performance score between 'y_true' and 'y_predict'
    score =  r2_score(y_true, y_predict)
# Return the score
    return score

In [186]:
# Calculate the performance of this model
score = performance_metric([3, -0.5, 2, 7, 4.2], [2.5, 0.0, 2.1, 7.8, 5.3])
print ("Model has a coefficient of determination, R^2, of {:.3f}.".format(score))

Model has a coefficient of determination, R^2, of 0.923.


## Question 2 answer 
 this model to have successfully captured the variation of the target variable
 
*** cause coefficient of determination (R^2) = .923 it means that 92.3% percent of the variance in Y_predict is predictable from 
the independent variables it close to 100% so it's a good prediction ratio 


## Implementation: Shuffle and Split Data

In [201]:
# Import 'train_test_split'
from sklearn.model_selection import train_test_split

# Shuffle and split the data into training and testing subsets
#Split the data into 80% training and 20% testing.
#Assign the train and testing splits to X_train, X_test, y_train, and y_test.
# random_state=0 to ensure that  results are consistent
X_train, X_test, y_train, y_test = train_test_split(features, prices,train_size=.8, test_size = 0.2, random_state=0)

# Success
print ("Training and testing split was successful.")

Training and testing split was successful.


## Question 3 answer 
 the benefit to splitting a dataset into some ratio of training and testing subsets for a learning algorithm :
 
 
* to know how well my model doing and compare between algorthims.
* to choose the best algorthim which achive the best accuracy based on testing and training sets.
* trying to reduce training time. 

## Analyzing Model Performance
In this third section of the project, you'll take a look at several models' learning and testing performances on various subsets of training data. Additionally, you'll investigate one particular algorithm with an increasing 'max_depth' parameter on the full training set to observe how model complexity affects performance. Graphing your model's performance based on varying criteria can be beneficial in the analysis process, such as visualizing behavior that may not have been apparent from the results alone.

Learning Curves
The following code cell produces four graphs for a decision tree model with different maximum depths. Each graph visualizes the learning curves of the model for both training and testing as the size of the training set is increased. Note that the shaded region of a learning curve denotes the uncertainty of that curve (measured as the standard deviation). The model is scored on both the training and testing sets using R2, the coefficient of determination.

Run the code cell below and use these graphs to answer the following question.

In [202]:
# Produce learning curves for varying training set sizes and maximum depths
vs.ModelLearning(features, prices)

AttributeError: module 'visuals' has no attribute 'ModelLearning'